File: c++_implementation/c++_implementation/merge.cpp
#include "merge.hpp"


Image mergeImages(Image im1, Image im2, double alpha) {
   

    int mergedW = std::max(im1.width, im2.width);
    int mergedH = std::max(im1.height, im2.height);
    Image* merged = new Image(mergedW, mergedH, 3);

    int size1 = im1.width * im1.height * 3;
    int size2 = im2.width * im2.height * 3;


    if(size1 > size2) {
        // Copy first image to output image
        for (int y = 0; y < im1.height; y++) {
            for (int x = 0; x < im1.width; x++) {
                int srcIndex = (y * im1.width + x) * 3;
                int destIndex = (y * mergedW + x) * 3;
                for (int c = 0; c < 3; c++) {
                    merged->imageData[destIndex + c] = im1.imageData[srcIndex + c];
                }
            }
        }
        // Merge images with proper resolution
        for (int y = 0; y < im2.height; y++) {
            for (int x = 0; x < im2.width; x++) {
                int srcIndex = (y * im2.width + x) * 3;
                int destIndex = (y * mergedW + x) * 3;
                for (int c = 0; c < 3; c++) {
                    merged->imageData[destIndex + c] = ((im1.imageData[destIndex + c] * (1.0 - alpha)) + (im2.imageData[srcIndex + c] * alpha));
                }
            }
        }
    }

    else {
        // Kopiuj zawartoœæ pierwszego obrazu do obrazu wynikowego
        for (int y = 0; y < im2.height; y++) {
            for (int x = 0; x < im2.width; x++) {
                int srcIndex = (y * im2.width + x) * 3;
                int destIndex = (y * mergedW + x) * 3;
                for (int c = 0; c < 3; c++) {
                    merged->imageData[destIndex + c] = im2.imageData[srcIndex + c];
                }
            }
        }
        for (int y = 0; y < im1.height; y++) {
            for (int x = 0; x < im1.width; x++) {
                int srcIndex = (y * im1.width + x) * 3;
                int destIndex = (y * mergedW + x) * 3;
                for (int c = 0; c < 3; c++) {
                    merged->imageData[destIndex + c] = ((im2.imageData[destIndex + c] * (1.0 - alpha)) + (im1.imageData[srcIndex + c] * alpha));
                }
            }
        }
    }
    
    return *merged;
}

File: python_implementation/merge.py
from PIL import Image


def mergeImages(im1, im2, alpha):
    mergedW = max(im1.width, im2.width)
    mergedH = max(im1.height, im2.height)

    merged = Image.new("RGB", (mergedW, mergedH))

    for y in range(mergedH):
        for x in range(mergedW):
            p1 = im1.getpixel((x * im1.width // mergedW, y * im1.height // mergedH))
            p2 = im2.getpixel((x * im2.width // mergedW, y * im2.height // mergedH))

            p1 = tuple(map(int, p1))
            p2 = tuple(map(int, p2))

            merged_pixel_value = tuple(
                int(p1_channel * (1.0 - float(alpha)) + p2_channel * float(alpha))
                for p1_channel, p2_channel in zip(p1, p2)
            )

            merged.putpixel((x, y), merged_pixel_value)

    return merged

